examples/refactoring.py

```diff
@@ -0,0 +1,168 @@
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+'''Refactored Example to Minimize code multiplication'''
+
+from neurom.core.tree import isegment, isection
+import neurom.analysis.morphtree as mtr
+
+from neurom.core.types import TreeType
+
+from neurom.core.types import checkTreeType
+
+from itertools import chain, imap
+
+
+class Neurite(object):
+    ''' Neurite Class
+    '''
+    def __init__(self, tree):
+        ''' Construct Neurite object
+        '''
+        self._tree = tree
+        self._type = mtr.get_tree_type(tree)
+
+    def segments(self):
+        '''Returns segment iterator
+        '''
+        return isegment(self._tree)
+
+    def sections(self):
+        '''Returns sections iterator
+        '''
+        return isection(self._tree)
+
+    @property
+    def type(self):
+        ''' Returns neurite type
+        '''
+        return self._type
+
+
+class Neuron(object):
+    ''' Neuron Class
+    '''
+    def __init__(self, soma, neurites, name='Neuron'):
+        '''Construct a Neuron
+
+        Arguments:
+            soma: soma object
+            neurites: iterable of neurite tree structures.
+            name: Optional name for this Neuron.
+        '''
+        self.soma = soma
+        self.neurites = neurites
+        self.name = name
+
+    def _i_neurites(self, func, tree_filter=None):
+        '''Returns a mapped iterator to all the neuron's neurites
+
+        Provides access to all the elements of all the neurites
+        in one iteration sequence.
+
+        Parameters:
+            iterator_type: type of the iteration (segment, section, triplet...)
+            mapping: optional function to apply to the iterator's target.
+            tree_filter: optional top level filter on properties of neurite tree objects.
+        '''
+        nrt = (self.neurites if tree_filter is None else filter(tree_filter, self.neurites))
+        return chain(*imap(func, nrt))
+
+    def segments(self, neurite_type=TreeType.all):
+        '''Returns segments
+        '''
+        return self._i_neurites(lambda n: n.segments(),
+                                tree_filter=lambda t: checkTreeType(neurite_type, t.type))
+
+    def sections(self, neurite_type=TreeType.all):
+        '''Returns sections
+        '''
+        return self._i_neurites(lambda n: n.sections(),
+                                tree_filter=lambda t: checkTreeType(neurite_type, t.type))
+
+
+class Population(object):
```

Comment:
I still don't understand why the population has to have different code to the neuron when it comes to getting any neurite related information. Both can store collections of neurites, so both could use exactly the same code.